### PR TITLE
feat(frontend-compiler-plugin): identify wrong use of pattern `evolve` + `neighboring` instead of using `share`

### LINKED ISSUE #575: Frontend compiler plugin should throw a warning for the possible wrong use of constructs.
The following use of the `repeat`( /`repeating` ) construct with a nested `neighboring`, in some cases should be highlighted as a warning from the frontend compiler plugin.

Eg:
``` kotlin
repeat(initial){ it ->
    val newValue = //same of `it`
    neighboring(newValue).operation //field reduction
}
```
In the case above, 
the compiler should throw a warning that suggest to use the `share` construct instead of nesting `repeat` and `neighboring`.
If what you share with your neighbors (using `neighboring`), and it depends on the state of the `repeat`, 
is different from what you return as the last value of the `repeat`, then the entire block can be replaced with a `share`.

``` kotlin
repeat(initial){ it ->
    val newValue = //same of `it`
    neighboring(otherValue)
    newValue
}
```
Different is the case in which a different value is shared among neighbors, in this case it is ok to use the nested constructs.

You may also find a pattern for this use.

### PR DESCRIPTION
**Pull request by FreshMag**
closes #575

### CODE REVIEW

File: compiler-plugin-test/src/test/resources/subjekt/ImproperConstruct.yaml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      - |-
+        evolve(0) { value ->
+            neighboring(value)
+            value
+        }
+      - |-
+        evolving(0) { value ->
+            neighboring(value)
+            value.yielding { "example" }
+        }
```
Comment (angelacorte) [inline compiler-plugin-test/src/test/resources/subjekt/ImproperConstruct.yaml:97]: what are you checking in these cases? the effective use of a value inside the `evolve`?  to be 100% fair, the correct use of the `evolve` should be the one in the `PROPER_CONSTRUCT_IMPLICIT/EXPLICIT`, since you should modify the evolving value, otherwise it has no sense to use the `evolve` ̇

Comment (FreshMag) [reply compiler-plugin-test/src/test/resources/subjekt/ImproperConstruct.yaml:97]: You’re absolutely right. Those two test cases are more of a “stress test” for the checker’s boundaries. Since the issue description highlights the necessary difference between the expression inside `neighboring` and the one it returns, I wanted to ensure that the checker doesn’t raise any warnings when the expressions are identical.

If you have any suggestions for defining this test case in a more realistic scenario, please let me know, and I’ll replace those code snippets accordingly.

Comment (angelacorte) [reply compiler-plugin-test/src/test/resources/subjekt/ImproperConstruct.yaml:97]: I'll think about other test cases and let you know if something comes up to my mind!

**Review by angelacorte — APPROVED** [review]

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            val expressionToCheck = neighboringExpression as FirExpression // needed for null safety
+            if (constructNameFQName == EVOLVING_FUNCTION_FQ_NAME) {
+                // if the construct is the `evolving` one we check if the yielding receiver is different from the
+                // neighboring one
+                val yieldingReceiver =
+                    with(YieldingReceiverVisitor()) {
+                        returnExpression.getYieldingReceiver()
+                    }
+
+                isReplaceable = yieldingReceiver?.isStructurallyEquivalentTo(expressionToCheck) == false
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt:140]:
```suggestion
                isReplaceable = !yieldingReceiver?.isStructurallyEquivalentTo(expressionToCheck)
```

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            } else {
+                isReplaceable = returnExpression.result.isStructurallyEquivalentTo(expressionToCheck) == false
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt:142]:
```suggestion
                isReplaceable = !returnExpression.result.isStructurallyEquivalentTo(expressionToCheck)
```

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            markExpression = false
+        }
+    }
+
+    override fun visitResolvedNamedReference(resolvedNamedReference: FirResolvedNamedReference) {
+        // mark the expression if it uses the parameters of the construct or the traced dependent symbols
+        if (resolvedNamedReference.resolvedSymbol in parametersDeclarations ||
+            resolvedNamedReference.resolvedSymbol in tracedDependentSymbols
+        ) {
+            markExpression = true
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt:90]: Why not simply replace the assignment with the if condition? Can it cause problems?

File: compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/KtDefaultErrorMessagesCollektive.kt [hunk 5]
```diff
@@ -56,5 +56,12 @@ object KtDefaultErrorMessagesCollektive : BaseDiagnosticRendererFactory() {
                 """.trimIndent(),
                 CommonRenderers.STRING,
             )
+            put(
+                FirCollektiveErrors.IMPROPER_EVOLVE_CONSTRUCT,
+                """
+                The ''{0}'' construct can be replaced with a simpler `share` construct call.
```
Comment (nicolasfara) [inline compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/checkers/KtDefaultErrorMessagesCollektive.kt:62]:
```suggestion
                The ''{0}'' construct can be replaced with the `share` construct call. 
```

**Review by nicolasfara — CHANGES_REQUESTED** [review]

**Review by nicolasfara — APPROVED** [review]
